mesh/mesh_operators.hpp

```diff
@@ -307,6 +307,105 @@ class ThresholdDerefiner : public MeshOperator
    virtual void Reset() { estimator.Reset(); }
 };
 
+/** @brief Refinement operator to control data oscillation.
+
+    This class computes osc_K(f) := || h ⋅ (I - Π) f ||_K  at each element K.
+    Here, Π is the L2-projection and ||⋅||_K is the L2-norm, restricted to the
+    element K. All elements satisfying the inequality
+    \code
+       osc_K(f) > threshold ⋅ ||f|| / sqrt(n_el),
+    \endcode
+    are refined. Here, threshold is a postive parameter, ||⋅|| is the L2-norm
+    over the entire domain Ω, and n_el is the number of elements in the mesh.
+
+    Note that if osc(f) = threshold ⋅ ||f|| / sqrt(n_el) for each K, then
+    \code
+       osc(f) = sqrt( sum_K osc_K^2(f)) = threshold ⋅ ||f||.
+    \endcode
+    This is the reason for the 1/sqrt(n_el) factor. */
+class CoefficientRefiner : public MeshOperator
+{
+protected:
+   int nc_limit = 1;
+   int nonconforming = -1;
+   int order;
+   long max_elements = std::numeric_limits<long>::max();
+   double threshold = 1.0e-2;
+   double global_osc = 0.0;
+   Array<int> mesh_refinements;
+   Vector element_oscs;
+   Coefficient *coeff = NULL;
+   GridFunction *gf;
+   const IntegrationRule *ir_default[Geometry::NumGeom];
+   const IntegrationRule **irs = NULL;
+
+   /** @brief Apply the operator to the mesh once.
+       @return STOP if a stopping criterion is satisfied or no elements were
+       marked for refinement; REFINED + CONTINUE otherwise. */
+   virtual int ApplyImpl(Mesh &mesh);
+
+public:
+   /// Constructor
+   CoefficientRefiner(int order_) : order(order_) { }
+
+   /** @brief Apply the operator to the mesh max_it times or until tolerance
+    *  achieved.
+       @return STOP if a stopping criterion is satisfied or no elements were
+       marked for refinement; REFINED + CONTINUE otherwise. */
+   virtual int PreprocessMesh(Mesh &mesh, int max_it);
+
+   int PreprocessMesh(Mesh &mesh)
+   {
+      int max_it = 10;
+      return PreprocessMesh(mesh, max_it);
+   }
+
+   /// Set the refinement threshold. The default value is 1.0e-3.
+   void SetThreshold(double threshold_) { threshold = threshold_; }
+
+   /** @brief Set the maximum number of elements stopping criterion: stop when
+       the input mesh has num_elements >= max_elem. The default value is
+       LONG_MAX. */
+   void SetMaxElements(long max_elements_) { max_elements = max_elements_; }
+
+   /// Set the function f
+   void SetCoefficient(Coefficient &coeff_)
```

Comment:
The coefficient seems to be required. At at minimum this requirement should be mentioned in the comments. Alternatively a `Coefficient &` argument could be added to the constructor. Yet another alternative would be to remove the requirement by using a constant value of 1 when the user doesn't provide a coefficient. I'd like to see more than just a change in the comment but I'm not certain what functional change would be most appropriate.